internal/controller/supervisorstorage/garbage_collector.go

```diff
@@ -121,28 +121,40 @@ func (c *garbageCollectorController) Sync(ctx controllerlib.Context) error {
 			continue
 		}
 
-		if garbageCollectAfterTime.Before(frozenClock.Now()) {
-			storageType, isSessionStorage := secret.Labels[crud.SecretLabelKey]
-			if isSessionStorage {
-				err := c.maybeRevokeUpstreamOIDCRefreshToken(ctx.Context, storageType, secret)
-				if err != nil {
-					// Log the error for debugging purposes, but still carry on to delete the Secret despite the error.
-					plog.WarningErr("garbage collector could not revoke upstream refresh token", err, logKV(secret))
-				}
-			}
+		if !garbageCollectAfterTime.Before(frozenClock.Now()) {
+			// not old enough yet
+			continue
+		}
 
-			err = c.kubeClient.CoreV1().Secrets(secret.Namespace).Delete(ctx.Context, secret.Name, metav1.DeleteOptions{
-				Preconditions: &metav1.Preconditions{
-					UID:             &secret.UID,
-					ResourceVersion: &secret.ResourceVersion,
-				},
-			})
+		storageType, isSessionStorage := secret.Labels[crud.SecretLabelKey]
+		if isSessionStorage {
+			err := c.maybeRevokeUpstreamOIDCRefreshToken(ctx.Context, storageType, secret)
 			if err != nil {
-				plog.WarningErr("failed to garbage collect resource", err, logKV(secret))
-				continue
+				// Log the error for debugging purposes, but still carry on to delete the Secret despite the error.
+				plog.WarningErr("garbage collector could not revoke upstream refresh token", err, logKV(secret))
+				// If the error is of a type that is worth retrying, then not delete the Secret right away.
+				// A future call to Sync will try revocation again for that secret. However, if the Secret is
+				// getting too old, then just delete it anyway. We don't want to extend the lifetime of these
+				// session Secrets by too much time, since the garbage collector is the only thing that is
+				// cleaning them out of etcd storage.
+				fourHoursAgo := frozenClock.Now().Add(-4 * time.Hour)
+				if errors.As(err, &retryableRevocationError{}) && garbageCollectAfterTime.After(fourHoursAgo) {
+					continue
```

Comment:
The math here is painful.  I do not think it can be improved though.

Reply:
Agreed. Adding another local variable and another comment to make it (a little) more readable.